### PR TITLE
Jackie - oo-ride-share - Ampers

### PR DESCRIPTION
**Pull request by Jackiesan**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I had to decide whether I wanted to assign drivers within the request_trip(passenger_id) method or whether it should be it's own method. Initially was assigning drivers in the same method as creating a new trip but since the method was starting to handle 2 responsibilities, I decided to make assign_driver a separate method within trip_dispatcher.rb. 
Describe any examples of composition that you encountered in this project, if any | I noticed that there were a lot of Has-A relationships. For example, Trip dispatcher has "trips", all trip objects have a "driver" and "passenger". Each passenger and driver also has the potential to have "trips".
Describe a _nominal test_ that you wrote for this assignment. | The test "returns drivers total revenue" in driver_spec #total_revenue method is a nominal test that asserts that the method will return the sum of the driver's earnings.
Describe an _edge case test_ that you wrote for this assignment. | The test "returns zero if no trips" within driver_spec #total_revenue method is an edge case. It's an edge case because there might be instances in which a driver has no trips. If a driver has no trips then there will be no trip costs (nil) and that can break the calculation. This test is assuring that this edge case is considered in the code.
Describe a concept that you gained more clarity on as you worked on this assignment. | I gained a lot more clarity on object oriented design. I understand now that a class can instantiate objects that are in different classes. I also learned that the class should only handle behaviors specific to that class. I'm also starting to view data more as behaviors.

### CODE REVIEW

File: lib/passenger.rb [hunk 1]
```diff
@@ -18,7 +19,36 @@ def get_drivers
...
     def add_trip(trip)
+      if trip.class != Trip
+        raise ArgumentError.new("Can only add trip instance to trip collection")
+      end
+
       @trips << trip
     end
+
+    def calculate_total_money_spent
+      if trips.length == 0
```
Comment (CheezItMan) [inline lib/passenger.rb:30]: Nice work checking different scenarios.  The 2nd gives a good argument for cost to return `nil` if the trip hasn't finished yet.

File: lib/trip_dispatcher.rb [hunk 2]
```diff
@@ -90,12 +92,74 @@ def load_trips
       trips
     end
 
+    def assign_driver
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:95]: Nicely done

File: specs/driver_spec.rb [hunk 3]
```diff
@@ -70,11 +82,176 @@
...
+        end_time: nil,
+        cost: nil,
+        rating: nil,
+      }
+
+      in_progress_trip = RideShare::Trip.new(in_progress_trip_data)
+
+      @driver.add_trip(in_progress_trip)
+
+      @driver.average_rating.must_equal 4.0
```
Comment (CheezItMan) [inline specs/driver_spec.rb:108]: Nicely done!

File: specs/driver_spec.rb [hunk 4]
```diff
@@ -70,11 +82,176 @@
...
+
+    end
+  end
+
+  describe "total_revenue method" do
```
Comment (CheezItMan) [inline specs/driver_spec.rb:113]: Well done, for the `total_revenue` method it looks like you handled the edge cases well!  🥇

File: specs/driver_spec.rb [hunk 5]
```diff
@@ -70,11 +82,176 @@
...
+
+    end
+
+    it "returns zero if no trips" do
+      driver = RideShare::Driver.new(id: 54, name: "Rogers Bartell IV", vin: "1C9EVBRM0YBC564DZ")
+      driver.total_revenue.must_equal 0
+    end
   end
+
+  describe "average revenue per hour method" do
```
Comment (CheezItMan) [inline specs/driver_spec.rb:179]: Ditto 🥇

File: specs/passenger_spec.rb [hunk 6]
```diff
@@ -75,4 +84,171 @@
...
+      start_time = Time.parse('2015-05-20T12:14:00+00:00')
+      end_time = start_time + 25 * 60
+
+      orginal_trips_count = @passenger.trips.length
+
+      trip = RideShare::Trip.new({id: 8, driver: RideShare::Driver.new(id: 3, name: "Lovelace", vin: "12345678912345678"), passenger: @passenger, start_time: start_time, end_time: end_time, rating: 5})
+
+      @passenger.add_trip(trip)
+
+      @passenger.trips.length.must_equal orginal_trips_count + 1
```
Comment (CheezItMan) [inline specs/passenger_spec.rb:103]: Having 
```ruby
@passenger.trips.must_include trip
```
would be appropriate as well.

File: specs/trip_dispatch_spec.rb [hunk 7]
```diff
@@ -88,5 +89,144 @@
...
+    it "stores end_time as instances of Time" do
+      dispatcher = RideShare::TripDispatcher.new
+
+      all_end_times = dispatcher.trips.all? { |trips| trips.end_time.class == Time }
+
+      all_end_times.must_equal true
+    end
+  end
+
+  describe "#assign_driver" do
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:112]: I am unconvinced that `assign_driver` needs to be a public method.  Why not make it `private`?

File: specs/trip_dispatch_spec.rb [hunk 8]
```diff
@@ -88,5 +89,144 @@
...
+        @new_trip.must_respond_to prop
+      end
+
+      @new_trip.id.must_be_kind_of Integer
+      @new_trip.driver.must_be_kind_of RideShare::Driver
+      @new_trip.passenger.must_be_kind_of RideShare::Passenger
+
+    end
+
+    it "raises error if there are no drivers available" do
```
Comment (CheezItMan) [inline specs/trip_dispatch_spec.rb:220]: Very well done

**Comment by CheezItMan** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Good number of commits and good commit messages, nice work!
Answer comprehension questions | Check, I'm glad design is making more sense.
**Wave 1** |
`Trip` has a helper method to calculate duration | Check
`Passenger` has a method to calculate total cost of all trips | Check
`Driver` has a method to calculate total revenue | Check
Complete tests | Very well done here
**Wave 2** |
`TripDispatcher` has a new method to create trips | Check
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | Check
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | Check
Complete tests | 
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | Check, but see my notes
Appropriate use of Ruby's `Date` | Check
Appropriate helper methods were made to help with complex logic | Check
Complete tests | Check
**Overall** | Very well done, this was outstanding work.  You hit all the learning goals and showed a really good understanding of testing and enumerables.
